### DISCUSSION TITLE
Set compilation target explicitly

### DISCUSSION THREAD

**Discussion by TanklesXL** [gleam-lang/gleam#1172] [Ideas & suggestions]
Now that `js` compilation support has been added to the gleam compiler, it might be useful to have some mechanism to optionally specify the intended compilation target of a gleam library/application, this would be useful to remove ambiguity when using/writing code that contains `external` function calls. 

There is probably a lot of nuance to this topic with regards to the benefits of being explicit v.s. just letting the compiler figure it out (as well as allowing for conditional compilation?), though I personally tend to favour explicitness when possible.

My suggestion would be to add a `target` field to `gleam.toml` with current options being  `erlang` (or  `erl`, ...), as well as  `js`. The way I see it, this field would only be required when there are any `external` functions present in the gleam code. 

In the case of any dependencies that would have to specify a target,  like in wrapper modules for erlang/elixir functionality, the target specified by all dependencies of a library would have to match, and would therefore force a given compilation target for the caller.

I'm sure plenty of other people have opinions/idea on this topic (especially those with more compiler knowledge than myself), and I look forward to seeing where this discussion goes.

Feel free to poke as many holes in my idea as possible :smile_cat:

**Comment by lpil** [comment]
This could be useful, I am curious about how it might impact tooling and general DX. What do you see as being the main experience for the user that is provided by this additional information?

I think we would want to be able to specify multiple supported target. For example the stdlib will support both Erlang and JavaScript.

What would the Gleam build tool do if it encounters a package marked as not for the current target? If it errors this may cause problem as packages that support both targets may have deps that are only used for one or the other. Alternatively they may need to provide wrapper 2 packages and a core package to work around this. Would this be limiting?

Would we record this on Hex in some way? They do have some amount of unstructured metadata.

**Reply by TanklesXL** [reply]
Hey @lpil  you make some good points, and I've given my original idea some further thought, so I have some more to run by you :smile: 

My main goal/motivation is to remove ambiguity, since we now have the option for multiple compilation targets, it is realistic to imagine that while ideally all gleam code **should** be compileable to both, in the case where it is using any `external` functionality, it would thereafter be limited to compiling for whichever ecosystem contains the external functionality being called.

This is however only a concern when the `external` functionality is actually being called, so if a dependency were to be 99% pure gleam and have a single optional-use external function wrapper, it would not be fair to limit the compilation target of the entire library. 

This brings me to my first new idea ( not sure how achievable it is/how much compiler work would be involved to do it but here goes :sweat_smile: ): 

Instead of having a `target` field in `gleam.toml`, it might be nice to bring the explicitness as close to the `external` functionality as possible, in this case to the external function signature. This could be achieved by adding a field in the signature that specifies the target, here is one example for each target, `console.log` for `js` and `string:length` for erlang. 

The benefit of this approach is that _in this case compilation target restrictions need only apply when the external function is actually used in the code being compiled_, this approach might actually enable external functionality for different targets to exist in the same module/library but ensure that they cannot both be used by the same calling module/library.

Old `js console.log` from [template-gleam-javascript](https://github.com/gleam-lang/template-gleam-javascript/blob/7d7ade950c9da30b99839819d9122b3b67de6533/src/main.gleam#L9):
```rust
external fn log(anything) -> Nil =
  "" "console.log"
```
New `js console.log`:
```rust
external fn log(anything) -> Nil =
  "js" "" "console.log"
```
or: 
```rust
external "js" fn log(anything) -> Nil =
 "" "console.log"
```

Old `erlang string:length` from [gleam_stdlib](https://github.com/gleam-lang/stdlib/blob/3e07ebbc721bd55d111c4e48af76dc8243e0149b/src/gleam/string.gleam#L48):
```rust
pub external fn length(String) -> Int =
  "string" "length"
```

New `erlang string:length`:
```rust
pub external fn length(String) -> Int =
  "erl" "string" "length"
```
or:
```rust
pub external "erl" fn length(String) -> Int =
  "string" "length"
```

It maybe becomes a little strange when it comes to using `Elixir` like for the call to `IO.inspect`, so maybe for the target name it might be worth using something like `beam` to refer to the whole erlang ecosystem instead of `erl`?
```rust 
pub external fn inspect(a) -> a = 
  "Elixir.IO" "inspect"
```
```rust 
pub external fn inspect(a) -> a = 
  "beam" "Elixir.IO" "inspect"
```

>Would we record this on Hex in some way?

For this I have no clue, I don't know anything about the internals of hex. :laughing:

**Reply by lpil** [reply]
> The benefit of this approach is that in this case compilation target restrictions need only apply when the external function is actually used in the code being compiled, this approach might actually enable external functionality for different targets to exist in the same module/library but ensure that they cannot both be used by the same calling module/library.

This is very interesting! I had 2 thoughts:

With this style how could we provide one external function for JS and another for Erlang?

Similarly, how would we provide an external function on one target and provide a Gleam implementation for the other?

It does remind me in a way of the tagging syntax in the conditional compilation issue, but I don't think that covered a way of marking code as not for a specific platform.

**Reply by TanklesXL** [reply]
> With this style how could we provide one external function for JS and another for Erlang?
Similarly, how would we provide an external function on one target and provide a Gleam implementation for the other?

I was wondering the same, perhaps if the functions have the same name & signature the compiler could be smart enough to decide which to use, though i know right now overloading is not supported so maybe that wouldn't work. 

*Note*: as far as i'm concerned any pure gleam function should always supercede the use of any external function. But in the case of the compiler deciding between 2 functions where 1 is gleam and the other external, the `external`one must be chosen _iff_ the gleam one makes a call at any point down the chain to  other `external` functionality for _a different target_.

The issue I have with allowing replaceable external functions is that there is no way to guarantee behavioural consistency even if the types coincide. For example even if 2 targets both support something like a `pub external fn to_string(a) -> String` function, we have no guarantees that given the same input, the call to both targets will provide the same output.

**Reply by TanklesXL** [reply]
Actually I get the feeling i might have misunderstood the question:
> With this style how could we provide one external function for JS and another for Erlang?

If you're asking how can separate external functionality for different targets live together, like having a `pub external "erl" fn a` and `pub external "js" fn b` in the same module and maintain proper compilation target guarantees, I'm not entirely sure cause I haven't thought about that part that much (hence why i said might :sweat_smile: )

**Reply by lpil** [reply]
> as far as i'm concerned any pure gleam function should always supercede the use of any external function. But in the case of the compiler deciding between 2 functions where 1 is gleam and the other external, the externalone must be chosen iff the gleam one makes a call at any point down the chain to other external functionality for a different target.

One place in which this wouldn't work is when we have a valid Gleam version but want to use a native version for performance or interop versions.

For example, we could implement `Map` in Gleam, but on Erlang we would want to use the native one as it will be both faster and fit better with the wider ecosystem. List reversing would be another example, for performance we use Erlang's version as it is written in C.

**Reply by TanklesXL** [reply]
Oh wow that's really cool, I had no idea that `reverse` was specially optimised, I knew it was external but i figured that was just to avoid redundancy/reinventing the wheel :open_mouth: 

it's an interesting consideration to make, i wonder what other languages like `Nim` that support multiple compile targets do in that case. Might there be a way of marking priority/precedence? or maybe only allow overloading/switching between `external` functions to avoid collisions/ambiguity?

**Reply by lpil** [reply]
I was thinking that we'd use conditional compilation to explicitly state which implemention to use for whatever platform, but maybe there's a better way?

```rust
if erlang {
  pub external fn reverse(List(a)) -> List(a) =
    "lists" "reverse"
} else {
  pub fn reverse(xs: List(a)) -> List(a) {
    todo
  }
}

**Comment by lpil** [comment]
This is done!
